Revisiting NpgsqlRange<T> boundaries and support for range literals

austindrenski opened the pull request:
# Overview

Range types in PostgreSQL are a breeze, but `NpgsqlRange<T>` can be complicated to work with. 

This PR is focused on improving `NpgsqlRange<T>` by:

 1. Simplifying construction
 2. Preventing bad states 
 3. Adding XML documentation
 4. Including a `TypeConverter`
 
## Simplifying construction

Currently, the fully specified constructor takes 6 arguments:

```c#
NpgsqlRange<int> range = new NpgsqlRange<int>(0, true, false, 5, true, false);
```

Contrast that with the range literal format in PostgreSQL:

```sql
SELECT '[0,5]'::int4range;
```

I've added a static method in https://github.com/npgsql/npgsql/commit/abbfe29174010e4ec36f0878c453ab7711ed3cd9 to parse the PostgreSQL range literal format:

```c#
NpgsqlRange<int> range = NpgsqlRange<int>.Parse("[0,5]");
```

I initially wrote `Parse` for a project that included `Microsoft.Extensions.Primitives.StringSegment`. I would prefer to keep using `StringSegment` here (nice API, reduce allocations), but it could be rewritten with low level `string`/`char[]` methods, if the added dependency is untenable...

## Preventing bad states

Currently, `NpgsqlRange<T>` uses private property setters via the constructors:
```c#
public bool LowerBoundIsInclusive
{
    get => (Flags & RangeFlags.LowerBoundInclusive) != 0;
    private set
    {
        if (value)
            Flags |= RangeFlags.LowerBoundInclusive;
       else
           Flags &= ~RangeFlags.LowerBoundInclusive;
    }
}
```

Given the arrival of `readonly struct`, I've attempted to rework the initialization logic to:
 1. Pass the initialization down to the `internal` constructor for consistency.
 2. Move the bitwise logic out of the setters and into a dedicated method.
 3. Add logic to guard against known bad states.
 4. Add recovery code for bad states that PostgreSQL would correct automatically (e.g. inclusive + infinite).
 5. Replaced the bitwise _checks_ with calls to `bool Enum.HasFlag(Enum)`.

Results:
```c#
/// <summary>
/// True if the lower bound is indefinite (i.e. infinite or unbounded); otherwise, false.
/// </summary>
public bool LowerBoundInfinite => Flags.HasFlag(RangeFlags.LowerBoundInfinite);
```

## Adding XML documentation

I've add some basic documentation to `NpgsqlRange<T>` and `RangeFlags`.

## Including a `TypeConverter`

I've added a `TypeConverter` as an internal nested class that calls up to the static `Parse(string)` method.

## Related items
https://github.com/npgsql/npgsql/issues/126#issue-24215665
https://github.com/npgsql/npgsql/issues/1418#issuecomment-280265251
https://github.com/npgsql/npgsql/pull/1493

roji commented on src/Npgsql/NpgsqlTypes/NpgsqlRange.cs:
```diff
@@ -131,49 +199,215 @@ internal NpgsqlRange([CanBeNull] T lowerBound, [CanBeNull] T upperBound, RangeFl
+                : (LowerBoundInfinite ? 0 : LowerBound.GetHashCode()) +
+                  (UpperBoundInfinite ? 0 : UpperBound.GetHashCode());
 
+        /// <inheritdoc />
```
Unrelated, but I just realized I replicated this logic in the EF Core provider's [`NpgsqlRangeTypeMapping.GenerateNonNullSqlLiteral()`](https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/blob/14d4579c59f42c3319cfa39fc0bcd0adca6a3c8a/src/EFCore.PG/Storage/Internal/Mapping/NpgsqlRangeTypeMapping.cs#L61). Do you want to compare the two and submit a PR which makes EF Core simply call `ToString()`?

austindrenski replied:
Sure, I'll take care of that soon.

austindrenski replied:
See: https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL/pull/420

roji commented on src/Npgsql/NpgsqlTypes/NpgsqlRange.cs:
```diff
@@ -131,49 +199,215 @@ internal NpgsqlRange([CanBeNull] T lowerBound, [CanBeNull] T upperBound, RangeFl
-            {
-                sb.Append(LowerBoundIsInclusive ? '[' : '(');
+
+            sb.Append(LowerBoundIsInclusive ? '[' : '(');
```
Any reason for making the distinction between inclusive/exclusive when the bound is infinite? The previous code just outputted `(` for infinity...

austindrenski replied:
This change was motivated by the [PostgreSQL docs](https://www.postgresql.org/docs/current/static/rangetypes.html#RANGETYPES-INFINITE):

> there is no such thing as an inclusive infinite bound — if you try to write one, it will automatically be converted to an exclusive bound.

Since the constructor rewrite converts inclusive-infinity to exclusive-infinity, the distinction here is no longer needed. The `string` representation of the lower bound for any initialized `NpgsqlRange<T>` will only depend on the inclusivity test, because all infinite bounds will be constructed as exclusive.

roji commented on src/Npgsql/NpgsqlTypes/NpgsqlRange.cs:
```diff
@@ -131,49 +199,215 @@ internal NpgsqlRange([CanBeNull] T lowerBound, [CanBeNull] T upperBound, RangeFl
+            if (value.Equals("empty", StringComparison.OrdinalIgnoreCase))
+                return Empty;
+
+            bool lowerInclusive = value[0] == '[';
```
nit: `var` everywhere please

roji commented on src/Npgsql/NpgsqlTypes/NpgsqlRange.cs:
```diff
@@ -131,49 +199,215 @@ internal NpgsqlRange([CanBeNull] T lowerBound, [CanBeNull] T upperBound, RangeFl
+            }
+
+            /// <inheritdoc />
+            public override object ConvertTo(ITypeDescriptorContext context, CultureInfo culture, object value, Type destinationType)
```
nit: use expression-bodied method

austindrenski commented on src/Npgsql/NpgsqlTypes/NpgsqlRange.cs:
```diff
@@ -19,109 +20,176 @@
+        /// <summary>
+        /// True if the lower bound is part of the range (i.e. inclusive); otherwise, false.
+        /// </summary>
+        public bool LowerBoundIsInclusive => Flags.HasFlag(RangeFlags.LowerBoundInclusive);
```
Ah, I wasn't thinking about the perf implications for the full framework. I'll revert the `Enum.HasFlag()` checks throughout the file.

Though, if those intrinsics are ever backported, it would be a nice readability improvement to use `Enum.HasFlag()`. I'll add a code comment citing this thread for posterity.

austindrenski commented on src/Npgsql/NpgsqlTypes/NpgsqlRange.cs:
```diff
@@ -19,109 +20,176 @@
 
-            LowerBoundIsInclusive = lowerBoundIsInclusive;
-            UpperBoundIsInclusive = upperBoundIsInclusive;
+            // TODO: We need to check if the bounds are implicitly empty. E.g. '(1,1)' or '(0,0]'.
```
I'm planning to handle this more thoroughly in a separate PR down the road. 

Basically, I think we need to incorporate the concept of range types being [continuous/discrete](https://www.postgresql.org/docs/current/static/rangetypes.html#RANGETYPES-DISCRETE) in order to handle this properly. 

I'm still thinking through the design, but one approach would be to [follow PostgreSQL](https://www.postgresql.org/docs/current/static/rangetypes.html#RANGETYPES-DEFINING) and have the user provide canonical functions, which would then allow us to increment/decrement for checks like this. I think the trick here would be to provide these hooks while guarding against obvious performance pitfalls.

roji replied:
OK, please open a separate issue for discussing this (you can copy-paste my comments below there).

Unless I'm mistaken, it's possible to identify discrete vs. continuous ranges in PostgreSQL by checking for the existence of a cannonicalization function ([`pg_range.rngcanonical`](https://www.postgresql.org/docs/current/static/catalog-pg-range.html)), as [described in the docs](https://www.postgresql.org/docs/current/static/rangetypes.html#RANGETYPES-DISCRETE).

The problem is that `NpgsqlRange<T>` isn't actually attached to any actual PostgreSQL database, and I think it's good to keep it that way. I think that a complex mechanism for loading discreteness from PostgreSQL would be a big overkill and of very little value (in theory two databases could also have the same range type defined but with different discreteness; but we really shouldn't go this far).

So it seems to me this is best handled by hardcoding the discreteness in code, which would probably cover at least 99% of the use cases.

austindrenski replied:
See: #1943

austindrenski commented:
Thanks for the review @roji , I pushed a commit that addresses the line-by-line comments and responded to those directly where you had specific questions. 

I'll address the general comments here:

> First, Span<char> is now supposed to provide more or less the same thing as StringSegment - allocation-free slicing. See also dotnet/corefx#20378 (comment) on StringSegment vs. ReadOnlySpan<char> - the latter is preferable, especially for our case where it really is a stack-only operation (no need to ever keep it in the heap).

This sounds great. I hadn't considered using the new `System.Memory` types for this, but I agree that they are a better fit.

> which by the way I think your PR omits from the csproj

I made some unrelated changes in the `csproj` file to get the project to build early on (before I knew this would be a PR). I'll revert and include that file if it ends up being relevant after sorting out the dependency question. 

> However, in 4.0.0 itself we don't use Span yet, so I'm leaning towards adding Span only for 4.1 - it would be weird to add a dependency just for NpgsqlRange<T>.Parse(), if you know what I mean. So I'm thinking it may make sense to implement it inefficiently for 4.0 and then do it with Span for 4.1.

I'm inclined to introduce the new dependency with the major version increment, given that there are plans to make extensive use of it throughout this major version.

That said, I think this falls squarely into the project/package management domain. If you find it sensible to hold back the dependency until its providing more value, I'm on board with that too.

roji commented on src/Npgsql/NpgsqlTypes/NpgsqlRange.cs:
```diff
@@ -131,49 +260,210 @@ internal NpgsqlRange([CanBeNull] T lowerBound, [CanBeNull] T upperBound, RangeFl
+                switch (value)
+                {
+                case string s:
+                {
```
These curlies are redundant, please remove.

Actually if you remove the `StringSegment` dependency the need for a switch may go away.

roji commented on test/Npgsql.Tests/Npgsql.Tests.csproj:
```diff
@@ -1,10 +1,11 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk" ToolsVersion="15.0">
   <PropertyGroup>
-    <TargetFrameworks>net451;netcoreapp2.0</TargetFrameworks>
+    <TargetFrameworks>netcoreapp2.0</TargetFrameworks>
```
As above, please revert to continue supporting net451 for now.

austindrenski replied:
Yikes, meant to be excluded from the commit. Removing.

roji commented on test/Npgsql.Tests/Types/RangeTests.cs:
```diff
@@ -190,5 +185,109 @@ public void OneTimeSetUp()
+        #region TheoryData
+
+        // ReSharper disable once InconsistentNaming
+        private static readonly DateTime May_17_2018 = DateTime.Parse("2018-05-17");
```
nit: please remove `private`

roji commented on src/Npgsql/Npgsql.csproj:
```diff
@@ -8,7 +7,7 @@
     <VersionPrefix>4.0.0</VersionPrefix>
     <LangVersion>latest</LangVersion>
-    <TargetFrameworks>net45;net451;netstandard2.0</TargetFrameworks>
+    <TargetFrameworks>netstandard2.0</TargetFrameworks>
```
I only now noticed that `Microsoft.Extensions.Primitives` doesn't support pre-netstandard2.0, which is another point against it in the context of Npgsql.

It's not that I care so much about supporting net45/net451 specifically, but continuing to support these platforms doesn't really cost us anything at the moment, so I wouldn't want to drop support for them just because of range string parsing...

austindrenski replied:
Same as above, meant to exclude that again.

roji commented on src/Npgsql/NpgsqlTypes/NpgsqlRange.cs:
```diff
@@ -19,109 +20,237 @@
+        /// <summary>
+        /// Defined by PostgreSQL to represent an empty range.
+        /// </summary>
+        const string EmptyLiteral = "empty";
```
I personally don't see huge value in having these strings as constants, but I don't have much against it either :) OK to leave like this if you prefer.

austindrenski replied:
I think there could be some value in documenting that these literals are based on PostgreSQL, rather than Npgsql. But I can use inline comments for the same effect. Let me know if you change your mind on keep/remove.

austindrenski commented:
> Let's leave this dependency out for now - I'd rather see everything out and stabilized before starting to adopt Span in the Npgsql codebase (which we'll definitely do).

Okay, I've removed the dependency on `Microsoft.Extensions.Primitives` and rewritten `NpgsqlRange<T>.Parse()` to work without it.

## Type constraints

I would like to constrain the type parameter with  `IComparable<T>` and `IEquatable<T>`. This will allow for more nuanced bounds checking to prevent additional error states. 

It seems reasonable to constrain the type parameter in this way as ranges are inherently modeling a comparable/equatable idea. Further, the PostgreSQL's built in range types are mapped to CLR types that implement these interfaces.

```c#
// before
public readonly struct NpgsqlRange<T> : IEquatable<NpgsqlRange<T>>
{ 
}

// after 
public readonly struct NpgsqlRange<T> : IEquatable<NpgsqlRange<T>> 
    where T : IComparable<T>, IEquatable<T>
{
}
```

## Additional bounds checks

With `IComparable<T>` and `IEquatable<T>`, it becomes possible to check for _some_ implicitly empty ranges, such as `[1, 1]`, though not the default valued range `[0, 0]`.

This also allows for the detection of malformed ranges where the lower bound is greater than the upper bound, such as `[1, 0]`.

## Bitwise flags

Since we're sticking with bitwise flag checks, I've added a few known states to the `RangeFlags` enum. The goal here is to make it more obvious when we're testing for a specific condition.

```c#
/// <summary>
/// Represents characteristics of range type boundaries.
/// </summary>
/// <remarks>
/// See: https://www.postgresql.org/docs/current/static/rangetypes.html
/// </remarks>
[Flags]
enum RangeFlags : byte
{
    ... // omitted

    /// <summary>
    /// Both the lower and upper bounds are inclusive.
    /// </summary>
    Inclusive = LowerBoundInclusive & UpperBoundInclusive,

    /// <summary>
    /// One of the bounds is inclusive and the other is exclusive.
    /// </summary>
    Mixed = LowerBoundInclusive ^ UpperBoundInclusive,

    /// <summary>
    /// The lower bound is both inclusive and infinite. This represents an error condition.
    /// </summary>
    LowerInclusiveInfinite = LowerBoundInclusive & LowerBoundInfinite,

    /// <summary>
    /// The upper bound is both inclusive and infinite. This represents an error condition.
    /// </summary>
    UpperInclusiveInfinite = UpperBoundInclusive & UpperBoundInfinite
}
```

roji commented:
> I would like to constrain the type parameter with IComparable<T> and IEquatable<T>. This will allow for more nuanced bounds checking to prevent additional error states.

Don't forget that in the end, PostgreSQL is the one responsible for making interpreting the range and raising errors. We can check for some simple error states client-side, but at the end of the day we send a range to PostgreSQL, which raises an error back if there's an issue. I'd like to avoid a complicated client-side validation which would add very little value (since validation already occurs server-side).

Another note is that at least theoretically, there may be a PostgreSQL range (possibly user-defined) over a type, whose C# analogue happens to not be equatable and/or comparable with itself. A good example may be `inet`: someone may want to have a range over `inet`, but .NET `IPAddress` happens to not be equatable or comparable. Adding the constraint would prevent users from creating the range, for little value (client-side validation/canonicalization which would be done server-side anyway).

Let me know what you think. For the rest I'm OK for merging, except that the PR doesn't currently build and imposes the equatable/comparable constraints.

austindrenski commented:
> I'd like to avoid a complicated client-side validation which would add very little value (since validation already occurs server-side).

I have a few projects that make use of `NpgsqlRange<T>`, and I usually end up defining a `MyRange<T>` type to wrap `NpgsqlRange<T>` when range data needs to leave (or hasn't yet reached) the data access layer.

It's certainly not Npgsql's responsibility to provide general purpose CLR types...but it would be nice to cut down on writing type-wrappers by incorporating more validation here. 

In any event, I've removed the `IComparable<T>` and `IEquatable<T>` constraints in the latest commit. The only anticipatable error state that we don't prevent is the `lower > upper` condition (which is easy enough to diagnose from consuming code, indeed making it of little value).

# Updates

I think I've completed the scope of feature changes for this PR. Along with removing those type constraints, the most recent commit also includes some housekeeping fixes and unit tests.

## Equality and hashes

I reworked the equality and hashcode methods to reflect the updated constructors. The unit tests all pass, but I would appreciate another set of eyes on the rewrite.

```c#
// old
public static bool operator ==(NpgsqlRange<T> x, NpgsqlRange<T> y)
    => x.IsEmpty == y.IsEmpty &&
       (x.LowerBoundInfinite || y.LowerBoundInfinite || x.LowerBound.Equals(y.LowerBound)) &&
       (x.UpperBoundInfinite || y.UpperBoundInfinite || x.UpperBound.Equals(y.UpperBound)) &&
       x.LowerBoundIsInclusive == y.LowerBoundIsInclusive &&
       x.UpperBoundIsInclusive == y.UpperBoundIsInclusive &&
       x.LowerBoundInfinite == y.LowerBoundInfinite &&
       x.UpperBoundInfinite == y.UpperBoundInfinite;



// new
public static bool operator ==(NpgsqlRange<T> x, NpgsqlRange<T> y)
    => x.Flags == y.Flags
       && (x.LowerBound?.Equals(y.LowerBound) ?? false)
       && (x.UpperBound?.Equals(y.UpperBound) ?? false);

public override int GetHashCode()
    => IsEmpty ? 0 : unchecked((397 * (int)Flags) ^ (397 * LowerBound?.GetHashCode() ?? 0) ^ (397 * UpperBound?.GetHashCode() ?? 0));
```

austindrenski commented:
I've removed the `IComparable<T>` and `IEquatable<T>` constraints in 4c4f96735335e387d47c047523d8546aa3cf6e67. 

At this point, I think the only anticipatable error state that we don't prevent is the `lower > upper` condition. Of all the possible error states, I think that one will be among the easiest for consuming code to get right.

# Updates

I think this commit completes the feature scope for this PR. Along with removing those type constraints, 4c4f96735335e387d47c047523d8546aa3cf6e67 also includes some housekeeping fixes and unit tests.

## Equality and hashes

I reworked the equality and hashcode methods to reflect the updated constructors. The unit tests all pass, but I would appreciate another set of eyes on the rewrite.

## Parser fixes

I've fixed a few bugs that were introduced in b55ed198934b5581e392c3cc442d79a632312879. I've also removed some of the bound-correction logic from the parser, so that bound-correction only ever occurs via the constructors.

## Unit tests

I've added a series of unit tests to parse various range literals. The tests use `NpgsqlRange<int>` and `NpgsqlRange<DateTime>`, but I can add others if there's any worry about subtype-specific errors.

austindrenski commented on src/Npgsql/NpgsqlTypes/NpgsqlRange.cs:
```diff
@@ -19,161 +20,509 @@
 
-            LowerBoundIsInclusive = lowerBoundIsInclusive;
-            UpperBoundIsInclusive = upperBoundIsInclusive;
+            if (lowerBound is T lower && !lower.Equals(default) && upperBound is T upper && !upper.Equals(default) && lower.Equals(upper))
```
`lowerBound is T lower` returns `false` when `lowerBound` is `null` (even when `T` is nullable). 

It's a nice shorthand around null checks. However, that line is removed in 156c5e55626bde38c9bc554c01fce711098afb95 as a result of addressing the boxing comment.

austindrenski commented:
> It should be possible to check if they happen to implement `IEquatable<T>`, and if so, call its `Equals()` method, avoiding the boxing. 

I moved the equality logic in 156c5e55626bde38c9bc554c01fce711098afb95 from `==` to `IEquatable<T>.Equals()`. It's a little easier (in my opinion) to reason about the equality logic with a clear primary entity (e.g. `this`).

I've added a static field `HasEquatableBounds` to avoid repeated type checks. At the same time, I've added another static field `HasValueTypeBounds` so that reference checks for `null` and value checks for `default(T)` are (hopefully) more transparent.

I've added a test that parses to `NpgsqlRange<int?>`, one that parses to `NpgsqlRange<string>`, and a third that parses to `NpgsqlRange<SimpleType>` which is just a test class to verify that everything works when `T` is neither a `ValueType` nor `IEquatable<T>`.

> If possible please run a BenchmarkDotNet to ensure there are zero allocations in this operation (the benchmark doesn't need to be checked in, just post its code and results).

I've never used BenchmarkDotNet, but I'll take a look now.

austindrenski commented:
> If possible please run a BenchmarkDotNet to ensure there are zero allocations in this operation (the benchmark doesn't need to be checked in, just post its code and results).

This is my first time using this library...so let me know if I've misconfigured something.

## Benchmark

```c#
[MemoryDiagnoser]
public class NpgsqlRangeEquality
{
    static readonly NpgsqlRange<int> A = NpgsqlRange<int>.Parse("(0,0)");
    static readonly NpgsqlRange<int> B = NpgsqlRange<int>.Empty;

    [Benchmark(Baseline = true)]
    public bool Equals() => A == B;
}
```

## Results

```
// * Summary *

.NET Core SDK=2.1.300-rc1-008673
  [Host]     : .NET Core 2.1.0-rc1 (CoreCLR 4.6.26426.02, CoreFX 4.6.26426.04), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.0-rc1 (CoreCLR 4.6.26426.02, CoreFX 4.6.26426.04), 64bit RyuJIT


 Method |     Mean |     Error |    StdDev | Scaled | Allocated |
------- |---------:|----------:|----------:|-------:|----------:|
 Equals | 3.767 ns | 0.1154 ns | 0.2766 ns |   1.00 |       0 B |

// * Hints *
Outliers
  NpgsqlRangeEquality.Equals: Default -> 1 outlier  was  removed

// * Legends *
  Mean      : Arithmetic mean of all measurements
  Error     : Half of 99.9% confidence interval
  StdDev    : Standard deviation of all measurements
  Scaled    : Mean(CurrentBenchmark) / Mean(BaselineBenchmark)
  Allocated : Allocated memory per single operation (managed only, inclusive, 1KB = 1024B)
  1 ns      : 1 Nanosecond (0.000000001 sec)

// * Diagnostic Output - MemoryDiagnoser *


// ***** BenchmarkRunner: End *****
// * Artifacts cleanup *

Global total time: 00:01:17 (77.8 sec)
```

roji commented on src/Npgsql/NpgsqlTypes/NpgsqlRange.cs:
```diff
@@ -19,161 +20,582 @@
+                    (((IEquatable<T>)UpperBound)?.Equals(other.UpperBound) ?? other.UpperBound == null);
 
-        public bool Equals(NpgsqlRange<T> other) => this == other;
+            return
```
Distinguishing between value types and references types isn't really necessary here. You can keep the code that checks for null, and it should get elided by the JIT.

roji commented:
Thanks @austindrenski, merged!

For the benchmarks, the most important part for us here is the fact that there are zero allocations. A few nanos here or there make very little difference - especially in a code-path which can be considered non-hot - but allocations quickly add up and have a perf impact on the application as a whole by creating more GC pressure.

One issue with your benchmark is that it compares an empty range, which means that the Equals method returns false immediately (at the flag comparison point), and does not actually check the important bit, which is casting to `IEquatable<T>` and calling equals. Can you please rerun it and post the results to confirm?

See also my comment about distinguishing between value and references types in the equality check.

I merged this in the meantime since 4.0 is right around the corner, you can still submit another PR on the above though.

roji commented:
One last comment - I try to follow [this guideline](https://chris.beams.io/posts/git-commit/) for commit messages, especially the 50-char and 72-char line limits, try to keep it in mind for future work.

And thanks for all the contributions!
